### PR TITLE
Text refactor

### PR DESCRIPTION
**Pull request by rfuest**
Thank you for helping out with embedded-graphics development! Please:

- [ ] Check that you've added passing tests and documentation
- [ ] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) if your changes affect the **public API**
- [ ] Run `rustfmt` on the project
- [ ] Run `just build` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

This PR adds support for external text renderers to e-g. `TextStyle` was converted to a trait that can be also implemented in other crates. To make drawing more efficient text renderers get direct access to the draw target and can use optimizations like `fill_contiguous`. Renderers can implement `TextStylePixels` to also provide a pixel iterator, but this isn't required.

Are the names `MonospacedTextStyle` and `MonospacedTextStyleBuilder` too long? Do you have a better suggestion?

### CODE REVIEW

**Comment by jamwaffles** [comment]
I'm a bit frazzled at the moment so I'll get round to a review later (sorry), but as for 

> Are the names MonospacedTextStyle and MonospacedTextStyleBuilder too long? Do you have a better suggestion?

How about `MonoTextStyle` and `MonoTextStyleBuilder`? You could change `MonospacedFont` to `MonoFont`, etc as well. `Mono` might be a bit ambiguous at this point, but my hope is that if it's used everywhere for all the monospaced fonts, the name along with all the documentation around it will be clear.

**Comment by rfuest** [comment]
I've also implemented an example of an external text renderer in my fork of the `simulator` repo: https://github.com/rfuest/simulator/blob/text-refactor/examples/text-seven-segment.rs

It can be used to draw seven segment displays with any scale:

grafik

The code is too complex to keep it as an example, but I think it could be useful to release the renderer as a separate crate.

**Comment by rfuest** [comment]
> I'm a bit frazzled at the moment so I'll get round to a review later (sorry), but as for

Never mind, it's only a draft PR and you can review it when it's done.

> How about `MonoTextStyle` and `MonoTextStyleBuilder`? You could change `MonospacedFont` to `MonoFont`, etc as well. `Mono` might be a bit ambiguous at this point, but my hope is that if it's used everywhere for all the monospaced fonts, the name along with all the documentation around it will be clear.

Sounds good to me, thanks for the suggestion.

File: src/fonts/mod.rs
Comment (bugadani) [inline src/fonts/mod.rs:211]: If you're at it, could you require MonoFont to implement Copy?

File: src/fonts/monospaced_pixels.rs
Comment (bugadani) [inline src/fonts/monospaced_pixels.rs:79]: Hmm, I already closed #439 assuming you'd done that. Would you incorporate those changes, or should I reopen it after this is merged?

File: src/fonts/monospaced_pixels.rs
Comment (bugadani) [inline src/fonts/monospaced_pixels.rs:74]: While it's not a huge difference, this lookup has a bit of perf impact. In embedded-text, I made a glyph renderer struct that contains the pixel offset instead of the character and avoids some unnecessary re-calculations for each pixel. Also I think it would be nice if this code could be reused in embedded-text, but I think I might implement that after this is merged :)

File: src/fonts/monospaced_pixels.rs
Comment (rfuest) [reply src/fonts/monospaced_pixels.rs:79]: I would suggest to merge #439 first and I'll rebase this PR to incorporate the changes in that PR.

File: src/fonts/mod.rs
Comment (rfuest) [reply src/fonts/mod.rs:211]: I think that should be possible, but I'll need to take a closer look.

File: src/fonts/monospaced_pixels.rs
Comment (rfuest) [reply src/fonts/monospaced_pixels.rs:74]: I want to try to reuse the `RawDataIter` that is used to draw images. This would reduce some code duplication but I'm not sure about the performance impact yet. If that doesn't work out we can look into optimizing this code.

File: src/fonts/mod.rs
Comment (rfuest) [reply src/fonts/mod.rs]: Done.

File: CHANGELOG.md
Comment (jamwaffles) [inline CHANGELOG.md:21]: So we can keep track of what needs to go in the migration guide instead of doing it all at once, could you add a section about these changes? Bullet points are fine if you prefer - I can flesh them out later.

File: src/fonts/monospaced_pixels.rs
Comment (jamwaffles) [reply src/fonts/monospaced_pixels.rs:74]: > While it's not a huge difference, this lookup has a bit of perf impact.

Did you see a perf impact running `just bench font`? I realise they're contrived benchmarks and perhaps not representative of realworld performance, but the text benches improve ~10% for me on this branch compared to `master`.

File: src/fonts/mod.rs
Comment (jamwaffles) [inline src/fonts/mod.rs:114]: FYI: I've opened #479 to track changes around fixing these links.

File: src/fonts/mod.rs
Comment (jamwaffles) [inline src/fonts/mod.rs:170]:
```suggestion
    /// Returns the position of a character in the font.
```

File: src/fonts/monospaced_pixels.rs
Comment (jamwaffles) [inline src/fonts/monospaced_pixels.rs:54]: Thanks for keeping this in :)

File: src/style/mono_text_style.rs
Comment (jamwaffles) [inline src/style/mono_text_style.rs:138]:
```suggestion
/// `MonoTextStyle` returned by `.build()`. This example draws white text with no background at
```

**Review by jamwaffles — CHANGES_REQUESTED** [review]
My eyes have gone a bit square looking at the word `Mono` over and over again, but looks good overall.

One thing that's missing that I think is important is an example of implementing a custom font. I don't know if it's possible to add one that's short enough to show in the docs, but it would really help tie all the traits and structs together.

There's a reference to the `text-custom-font` simulator example somewhere (I can't find it in the diff now for some reason) which might be worth turning into a link too.

File: src/fonts/monospaced_pixels.rs
Comment (rfuest) [reply src/fonts/monospaced_pixels.rs:54]: Sorry, but the final verdict is still open on this one. After this PR each font renderer still needs to handle newlines itself, but I'll look into it after this is merged.

File: CHANGELOG.md
Comment (rfuest) [reply CHANGELOG.md:21]: I've added some very basic bullet points to the migration guide. I'll try to remember to update the guide when I add more text features in future PRs.

**Comment by rfuest** [comment]
I couldn't reply to this above:

> Did you see a perf impact running just bench font? I realise they're contrived benchmarks and perhaps not representative of realworld performance, but the text benches improve ~10% for me on this branch compared to master.

I hadn't noticed that before, but I also get some performance improvements. Not quite 10%, but I hadn't expected any performance gains at all, so the 3%-7% are also welcome.

Since the benches were updated to use a framebuffer they should be pretty representative, but only for displays drivers with a framebuffer. Perhaps we should look into also adding benches for other displays, where the number of draw calls or the amount of data that has to be transmitted are better metrics.

File: src/fonts/monospaced_pixels.rs
Comment (jamwaffles) [reply src/fonts/monospaced_pixels.rs:54]: I think that's ok for external fonts. I'd like to keep at as a feature of the builtin mono fonts if it's not too complicated. Otherwise I'll concede and just use multiple `Text` calls :)

Comment (rfuest) [reply src/fonts/monospaced_pixels.rs:54]: > I think that's ok for external fonts.

I don't ;). I would like to keep the external text renderer API as simple as possible.

> Otherwise I'll concede and just use multiple `Text` calls :)

I don't think you need to. It should be possible to move the handling of newlines from the text renderers to the styled text drawable. This way each text renderer would only need to render single lines and get the the multiline support for free. But I'll investigate this in another PR.

**Comment by rfuest** [comment]
> One thing that's missing that I think is important is an example of implementing a custom font. I don't know if it's possible to add one that's short enough to show in the docs, but it would really help tie all the traits and structs together.
> 
> There's a reference to the `text-custom-font` simulator example somewhere (I can't find it in the diff now for some reason) which might be worth turning into a link too.

An example in the docs wouldn't be much shorter than `text-custom-font`. If it's OK, I would like to create a new issue for the doc improvements and wait until the text refactor is finished.

File: src/fonts/monospaced_pixels.rs
Comment (jamwaffles) [reply src/fonts/monospaced_pixels.rs:54]: Sounds good!

**Comment by jamwaffles** [comment]
> An example in the docs wouldn't be much shorter than text-custom-font. If it's OK, I would like to create a new issue for the doc improvements and wait until the text refactor is finished.

I thought that might be the case. No worries - we can add more in later.

**Review by jamwaffles — APPROVED** [review]
Looks good from my side. I don't know if @bugadani had anything else to add?

**Comment by bugadani** [comment]
Don't wait for me, I can't review in depth :(

**Comment by rfuest** [comment]
> I've also implemented an example of an external text renderer in my fork of the `simulator` repo: https://github.com/rfuest/simulator/blob/text-refactor/examples/text-seven-segment.rs
> 
> It can be used to draw seven segment displays with any scale:
> 
> grafik
> 
> The code is too complex to keep it as an example, but I think it could be useful to release the renderer as a separate crate.

@jamwaffles  If I decide to convert this example into a reusable crate do you think it would be useful enough to be an official e-g project or should I release it myself?

**Comment by jamwaffles** [comment]
I think it'd be right at home in the e-g org if that's what you mean. Would it end up being published on crates.io? 

As complex as it is, it would also be a good example for those looking to implement a custom font.

**Comment by rfuest** [comment]
> I think it'd be right at home in the e-g org if that's what you mean.

Yes, that's what I meant.

> Would it end up being published on crates.io?

Yes. IMO a freely scalable seven segment font could be useful for different projects.

> As complex as it is, it would also be a good example for those looking to implement a custom font.

Just to be clear: It's not a custom font, but a custom text renderer. But I agree that it could still be a good example for how an external renderer can be integrated.

**Comment by jamwaffles** [comment]
> Yes, that's what I meant.

> Yes. IMO a freely scalable seven segment font could be useful for different projects.

Definitely! You should be able to make repositories in the e-g org, so feel free to do so when you have time. Let me know if you run into permissions errors.

> Just to be clear: It's not a custom font, but a custom text renderer. But I agree that it could still be a good example for how an external renderer can be integrated.

Right, renderer not font. I'll do my best to be clearer with the terminology from now on :)

**Comment by rfuest** [comment]
>  I'll do my best to be clearer with the terminology from now on :)

I just wanted to make sure we talk about the same thing ;).

**Pull request merged**
